File: bukkit/src/main/java/io/wdsj/asw/bukkit/util/VirtualThreadUtils.java
package io.wdsj.asw.bukkit.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Method;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;

/**
 * Utility class for virtual threads which are introduced in Java 21
 */
public class VirtualThreadUtils {
    private VirtualThreadUtils() {}

    private static ThreadFactory virtualThreadFactory;

    private static ExecutorService virtualThreadPerTaskExecutor;

    static {
        try {
            Method ofVirtual = Thread.class.getMethod("ofVirtual");
            Class<?> ThreadBuilder = Class.forName("java.lang.Thread$Builder");
            Method factory = ThreadBuilder.getMethod("factory");
            ofVirtual.setAccessible(true);
            factory.setAccessible(true);
            virtualThreadFactory = (ThreadFactory) factory.invoke(ofVirtual.invoke(null));
        } catch (Exception e) {
            virtualThreadFactory = null;
        }
        try {
            Method method = Executors.class.getMethod("newVirtualThreadPerTaskExecutor");
            method.setAccessible(true);
            virtualThreadPerTaskExecutor = (ExecutorService) method.invoke(null);
        } catch (Exception e) {
            virtualThreadPerTaskExecutor = null;
        }
    }

    @Nullable
    public static ThreadFactory newVirtualThreadFactory() {
        return virtualThreadFactory;
    }

    @Nullable
    public static ExecutorService newVirtualThreadPerTaskExecutor() {
        return virtualThreadPerTaskExecutor;
    }

    @NotNull
    public static ThreadFactory newVirtualThreadFactoryOrProvided(ThreadFactory threadFactory) {
        return Utils.checkNotNullWithFallback(virtualThreadFactory, threadFactory);
    }

    @NotNull
    public static ThreadFactory newVirtualThreadFactoryOrDefault() {
        return Utils.checkNotNullWithFallback(virtualThreadFactory, Executors.defaultThreadFactory());
    }

    @NotNull
    public static ExecutorService newVirtualThreadPerTaskExecutorOrProvided(ExecutorService executorService) {
        return Utils.checkNotNullWithFallback(virtualThreadPerTaskExecutor, executorService);
    }
}

File: bukkit/src/main/kotlin/io/wdsj/asw/bukkit/util/VirtualThreadUtils.kt
package io.wdsj.asw.bukkit.util

import java.lang.reflect.Method
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory

/**
 * Utility class for virtual threads which are introduced in Java 21
 */
object VirtualThreadUtils {
    private var methodVirtualThreadFactory: Method?
    private var methodThreadOfVirtual: Method?

    private var methodVirtualThreadPerTaskExecutor: Method?

    init {
        try {
            methodThreadOfVirtual = Thread::class.java.getMethod("ofVirtual")
            val threadBuilder = Class.forName("java.lang.Thread\$Builder")
            methodVirtualThreadFactory= threadBuilder.getMethod("factory")
            methodThreadOfVirtual?.isAccessible = true
            methodVirtualThreadFactory?.isAccessible = true
        } catch (e: Exception) {
            methodThreadOfVirtual = null
            methodVirtualThreadFactory = null
        }
        try {
            methodVirtualThreadPerTaskExecutor = Executors::class.java.getMethod("newVirtualThreadPerTaskExecutor")
            methodVirtualThreadPerTaskExecutor?.isAccessible = true
        } catch (e: Exception) {
            methodVirtualThreadPerTaskExecutor = null
        }
    }

    @JvmStatic
    fun newVirtualThreadFactory(): ThreadFactory? {
        return invokeOfVirtualFactory()
    }

    @JvmStatic
    fun newVirtualThreadPerTaskExecutor(): ExecutorService? {
        return invokeNewVirtualThreadPerTaskExecutor()
    }

    @JvmStatic
    fun newVirtualThreadFactoryOrProvided(threadFactory: ThreadFactory): ThreadFactory {
        return invokeOfVirtualFactory() ?: threadFactory
    }

    @JvmStatic
    fun newVirtualThreadFactoryOrDefault(): ThreadFactory {
        return invokeOfVirtualFactory() ?: Executors.defaultThreadFactory()
    }

    @JvmStatic
    fun newVirtualThreadPerTaskExecutorOrProvided(executorService: ExecutorService): ExecutorService {
        return invokeNewVirtualThreadPerTaskExecutor() ?: executorService
    }

    private fun invokeNewVirtualThreadPerTaskExecutor(): ExecutorService? {
        return try {
            methodVirtualThreadPerTaskExecutor?.invoke(null) as ExecutorService
        } catch (e: Exception) {
            null
        }
    }

    private fun invokeOfVirtualFactory(): ThreadFactory? {
        return try {
            methodVirtualThreadFactory?.invoke(methodThreadOfVirtual?.invoke(null)) as ThreadFactory
        } catch (e: Exception) {
            null
        }
    }
}
